[Translation] added extract messages from controllers

thomasage opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

Added a loop on controllers to extract messages.

Commit:
[Translation] added extract messages from controllers

iltar commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -127,6 +127,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
         if ($this->defaultViewsPath) {
             $viewsPaths[] = $this->defaultViewsPath;
         }
+        $controllersPaths = array($kernel->getRootDir().'/Controller');
```
What if my controllers are somewhere else?

thomasage replied:
@iltar you're right. I added a parameter to constructor (as for the views). But I'm not sure how to test it. Can you help me?

iltar replied:
What about making the parameter an array of paths? That would cover multiple locations. You can test it by writing fixtures for your test, writing controllers with translations.

thomasage replied:
An array of paths sounds great to me, but for the tests I wonder how to set path(s) when I call the command on CLI (bin/console translation:update)?

When I try your suggestion, it returns null for $this->router.
I'm not sure it's OK to have a behaviour for templates (parameter in constructor) and another behaviour for controllers (find auto).

Commit:
[Translation] added extract messages from controllers

linaori reviewed (changes requested):
There's a small issue with the current implementation:
 - It assumes a certain directory
 - I can only add 1 extra directory

It's hard to find all controllers (but not impossible).

In a command, you could do something like this:
```php
public function execute(InputInterface $input, OutputInterface $output)
{
    $rows = [];

    foreach ($this->router->getRouteCollection() as $route) {
        if (!$route->hasDefault('_controller')) {
            continue;
        }

        $request = new Request([], [], ['_controller' => $route->getDefault('_controller')]);
        $controller = $this->resolver->getController($request);
    }
}
```

The other option would be to pass a list of resolved controllers or a list of directories (less accurate).

Commit:
[Translation] added custom directories for controllers

iltar commented:
You'll have to inject the router `RouterInterface` (`router` service) for it to work.

Commit:
[Translation] fix coding style

thomasage commented:
Do you think it's a good option to have 2 different behaviours for templates and controllers?

thomasage commented:
If I inject RouterInterface, I will break BC (constructor signature will require a new parameter).

iltar commented:
You can make it optional, or not typehint it and resolve the actual value (+deprecation warning) in the constructor

Commit:
[Translation] added auto-discover of controllers (RouterInterface injection)

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -12,14 +12,15 @@
-use Symfony\Component\Console\Input\InputInterface;
-use Symfony\Component\Console\Output\OutputInterface;
-use Symfony\Component\Console\Input\InputArgument;
-use Symfony\Component\Console\Input\InputOption;
```
You should revert the reordering of use statements (even if this is correct, we do not do it as it makes merging old branches more complex).

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -62,19 +67,22 @@ public function __construct(TranslationWriterInterface $writer, TranslationReade
+                    new InputOption('no-backup', null, InputOption::VALUE_NONE, 'Should backup be disabled'),
+                    new InputOption('clean', null, InputOption::VALUE_NONE, 'Should clean not found messages'),
+                    new InputOption('domain', null, InputOption::VALUE_OPTIONAL, 'Specify the domain to update'),
+                )
```
You should also revert the CS change here to avoid making merging more complex with older branches.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -111,7 +118,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                    'Wrong output format',
+                    'Supported formats are: '.implode(', ', $supportedFormats).'.',
+                )
+            );
```
should be on one line.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -249,7 +287,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                'path' => $bundleTransPath,
+                'default_locale' => $this->defaultLocale,
+            )
+            );
```
should be on one line.

Commit:
FIX revert the reordering of use statements
FIX revert the CS change here to avoid making merging more complex with older branches
FIX one line instructions

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php:
```diff
@@ -12,11 +12,13 @@
 namespace Symfony\Bundle\FrameworkBundle\Tests\Command;
 
 use PHPUnit\Framework\TestCase;
+use Symfony\Bundle\FrameworkBundle\Command\TranslationUpdateCommand;
```
This one still needs to be reverted.

Commit:
FIX revert the reordering of use statements

nicolas-grekas commented:
Instead of hardcoding something for controllers only, what about allowing to configure a list of paths in a "translation.something" configuration option and make the command look there? That would work for controllers but also for anything else, wdyt?

Another idea could be to detect all services that have the translator injected into them and consider them all as targets for translation extraction.

thomasage commented:
> Instead of hardcoding something for controllers only, what about allowing to configure a list of paths in a "translation.something" configuration option and make the command look there? That would work for controllers but also for anything else, wdyt?
> 
> Another idea could be to detect all services that have the translator injected into them and consider them all as targets for translation extraction.

I prefer your 2nd idea, but I've no clue how do that. I don't think that "grep" every PHP classes is a right way?

yceruto commented:
See https://github.com/symfony/symfony/pull/29121 for alternative solution implementing the 2nd idea.

thomasage commented:
Closed in favor of #29121.
